Convert features and their tests to Typescript

Wihito opened the pull request:
So far I've converted most of the features and their tests, along with some lib files.

Wihito committed:
Convert features and their tests to Typescript (first draft)

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Fixup

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Use fake deck helper in edhrec suggestions test

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Use q param instead of query

crookedneighbor commented on src/js/types/deck.ts:
```diff
@@ -20,6 +20,7 @@ export interface Card {
     type_line?: string;
     oracle_id?: string;
   };
+  color_identity: color[] | ["C"];
```
FYI, this is not correct.

This type is for the deck entry object on the Scryfall deckbuilder, but `color_identity` is a property that exists on the Scryfall API response, not the deck entry object.

Wihito replied:
Okay, I removed that line and the color type now

crookedneighbor commented on test/features/card-input-modifier/index.test.ts:
```diff
@@ -1,20 +1,30 @@
-import mutation from "Lib/mutation";
+import * as bus from "framebus";
+import CardInputModifier from "../../../src/js/features/deck-builder-features/card-input-modifier";
+import deckParser from "../../../src/js/lib/deck-parser";
```
Any special reason these are not using the aliases? Maybe you were working on a branch before I added them to the tsconfig?

https://github.com/crookedneighbor/shambleshark/blob/master/tsconfig.json#L5-L16

Wihito replied:
I think so. I'll look through the files and adjust them, and see if my IDE still does these wonky imports.

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Fixup Main Feature class

crookedneighbor committed:
Merge branch 'fixups' into convert-features-and-their-tests-to-typescript

Wihito committed:
Convert features and their tests to Typescript (second draft)

crookedneighbor committed:
Remove invalid eslint rule

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Fixup Scryfall types

crookedneighbor committed:
Fix getCollection call

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Fixup tests

crookedneighbor committed:
Revert async behavior for cardin input modifier

crookedneighbor committed:
Move feature specific types to Feature class

crookedneighbor committed:
Simplify flatten token collection function per @Wihito

https://github.com/crookedneighbor/shambleshark/commit/afaa3b966d86f2b8513f22a0502dde1600377391#r39241673

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Simplify Feature classes

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Remove unused type

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Convert global features to ts

crookedneighbor committed:
convert edhrec suggestions to ts

crookedneighbor committed:
Update mutation lib to be a generic

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Add todo

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Fixup framebus

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Update mutation lib to use mocks pattern

crookedneighbor committed:
remove unnecessary type casting

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Finish converting edhrec suggestions to typescript

crookedneighbor committed:
Fix up edhrect suggestions

crookedneighbor committed:
Cleanup edhrec suggestions

crookedneighbor committed:
Move edhrec types to types directory

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Remove TODO comment

crookedneighbor committed:
Re-org tagger link

crookedneighbor committed:
Merge branch 'master' into convert-features-and-their-tests-to-typescript

crookedneighbor committed:
Finish up tagger link feature conversion

crookedneighbor committed:
Finish up token list

crookedneighbor committed:
Finish up scryfall search feature
